Better dirent access (reading)

mgautierfr opened the pull request:
This is the first PR corresponding to the split of #466
~~Based on #483 (not really needed for this PR but will be for the next, based on this one)~~ (merged in master)

- This change the architecture of the code as #466 does but do not change the creation behavior.
- Cherry-pick the `DirentReader` created by @veloman-yunkan in PR #430.
- ~~Add tests on zim creation.~~ (moved in #493)

veloman-yunkan commented on src/writer/creator.cpp:
```diff
@@ -220,14 +220,36 @@ namespace zim
-        auto dirent = handler->getDirent();
+        auto dirent = handler->getUniqueDirent();
         auto provider = handler->getContentProvider();
+        auto providerSize = provider->getSize();
```
`providerSize` is unused. Does `provider->getSize()` have a significant side-effect?

mgautierfr replied:
Code moved in #493

veloman-yunkan commented on src/writer/handler.h:
```diff
@@ -43,6 +43,18 @@ class Item;
+ * - Get the dirent associated to the handler using `getDirent()`.
+ *   If a handler has to handle itself, it has to do it itself before (in start/stop, ...)
+ *   The handlers will NOT have dirents of other hanlders passed.
+ *   TitleListingHandler is a exception, it will receive ALL dirents, including its own.
```
Therer is no `TitleListingHandler` yet

mgautierfr replied:
Code moved in #493

veloman-yunkan commented on src/writer/creator.cpp:
```diff
@@ -600,6 +600,7 @@ int mode =  _S_IREAD | _S_IWRITE;
               << " redirecting to (missing) "
               << dirent->getRedirectNs() << '/' << dirent->getRedirectPath());
           dirents.erase(dirent);
+          dirent->markRemoved();
```
This flow is not covered by unit tests

mgautierfr replied:
Code moved in #493

veloman-yunkan reviewed (changes requested):
1. The biggest changes brought by this PR are about dirent access and `zim::writer::DirentHandler`. Please change the title and description to emphasize and summarize that directly. (Frankly, I think that it would be better to further split it into two PRs separately focusing on the dirent access and `zim::writer::DirentHandler`, however I am afraid to say that :wink: ).

2. The message of the commit named "Readd cache system in DirentAccessor." refers to a previous commit by id. Better use the commit title, since the id is not automatically updated during a rebase.

mgautierfr commented on src/writer/creatordata.h:
```diff
@@ -121,6 +122,13 @@ namespace zim
         bool withIndex;
         std::string indexingLanguage;
+
+        std::vector<std::shared_ptr<DirentHandler>> m_handlers;
```
Code moved in #493

mgautierfr commented on src/writer/xapianHandler.cpp:
```diff
@@ -0,0 +1,106 @@
+
+FullTextXapianHandler::FullTextXapianHandler(CreatorData* data)
+  : mp_indexer(new XapianIndexer(data->basename+"_title.idx", data->indexingLanguage, IndexingMode::FULL, true)),
+    mp_creatorData(data)
```
Code moved in #493

mgautierfr commented on src/writer/xapianHandler.cpp:
```diff
@@ -0,0 +1,106 @@
+}
+
+TitleXapianHandler::TitleXapianHandler(CreatorData* data)
+  : mp_indexer(new XapianIndexer(data->basename+"_fulltext.idx", data->indexingLanguage, IndexingMode::TITLE, true)),
```
Code moved in #493

mgautierfr commented on src/writer/handler.h:
```diff
@@ -64,7 +69,11 @@ class DirentHandler {
     virtual void handle(Dirent* dirent, std::shared_ptr<Item> item) = 0;
 
   protected:
+    virtual Dirent* getDirent() const = 0;
```
Code moved in #493

mgautierfr commented on src/writer/creator.cpp:
```diff
@@ -198,35 +199,40 @@ namespace zim
-      data->addData('M', name, mimetype, std::move(provider), compressContent);
+      auto dirent = data->createDirent('M', name, mimetype, "");
+      data->addItemData(dirent, std::move(provider), compressContent);
+      data->handle(dirent);
```
Code moved in #493

mgautierfr commented:
> The biggest changes brought by this PR are about dirent access and zim::writer::DirentHandler. Please change the title and description to emphasize and summarize that directly. (Frankly, I think that it would be better to further split it into two PRs separately focusing on the dirent access and zim::writer::DirentHandler, however I am afraid to say that wink ).

I've split again this PR between the reading part (here) and the writing part (#493)

>  The message of the commit named "Readd cache system in DirentAccessor." refers to a previous commit by id. Better use the commit title, since the id is not automatically updated during a rebase.

Commit message changed.

mgautierfr commented on src/fileimpl.cpp:
```diff
@@ -155,7 +156,7 @@ sectionSubReader(const FileReader& zimReader, const std::string& sectionName,
+      result = std::min(result, mp_urlDirentAccessor->getOffset(entry_index_t(0)).v);
 
       // assuming that clusters are placed in the zim file in the same
       // order as the corresponding entries in the cluster pointer table
```
Yes. But :

- This is a low assumption (if the assumption is not true, the value may be not optimal but is still valid).
- This assumption is not new in this PR. If we have to fix it, let's fix it elsewhere. This PR is big enough :)

mgautierfr commented on src/fileimpl.cpp:
```diff
@@ -228,6 +228,18 @@ sectionSubReader(const FileReader& zimReader, const std::string& sectionName,
+    }
+    return title.compare(dirent.getTitle());
+  }
+
```
Done in fixup! commit.

veloman-yunkan reviewed (approved):
Thanks for splitting the PR again. I believe that moving forward with small and narrowly focused changes is beneficial for everyone involved.

kelson42 commented:
Unfortunately, codeFactor and codecov still complain. Would be better to get all green befoire merging.

mgautierfr commented:
- I've fixed the CodeFactor issues.
- I've removed some unused (not tested) methods. This increment a bit the coverage of the path.

What is not tested seems to be related to :
- Error handling (throwing ZimFileFormatError)
- Reading dirent > 256 bits (the patch change/move the code but do not change the algorithm)
- Handling deleted and target direct (never used as far as I know)

While I agree it would be nice to test the first two points, I think it is not related to the PR (it technically change the line but the code is not really changed)
I don't know what @veloman-yunkan thinks about this but I suggest we can merge this PR as it is (after a rebase-fixup).

veloman-yunkan commented:
> I don't know what @veloman-yunkan thinks about this but I suggest we can merge this PR as it is (after a rebase-fixup).

I concur with this point of view. In fact, before approving this PR I had checked the codecov/patch report and didn't see anything that could justify further delays.

mgautierfr commented:
Rebase-fixup done.
Ready to be merge once CI has run.
